### PR TITLE
[Feature][BatchNorm] Add BatchNormFwdOp and BatchNormBwdOp with TileLang kernels

### LINKED ISSUE #303: [FEAT][BATCH_NORM] Implement TileLang kernel and PyTorch-compatible Op for Batch Normalization
## Background

Batch Normalization (BN) is a foundational technique in deep neural networks, introduced in [Ioffe & Szegedy (2015)](https://arxiv.org/pdf/1502.03167). It normalizes activations across the batch dimension to reduce internal covariate shift, enabling higher learning rates and more stable training.

PyTorch exposes BN through a unified backend via `torch.nn.functional.batch_norm`, which powers:
- `torch.nn.BatchNorm1d` — input shape `(N, C)` or `(N, C, L)`
- `torch.nn.BatchNorm2d` — input shape `(N, C, H, W)`
- `torch.nn.BatchNorm3d` — input shape `(N, C, D, H, W)`

All three normalize over the `(N, *)` dimensions for each channel `C`.

TileOPs currently has no Batch Normalization implementation. This issue proposes adding a high-performance BN kernel and Op following the existing 2-layer architecture.

---

## Algorithm Summary

**Forward pass (training mode):**

```
# For each channel c:
mu_c    = mean(x[:, c, ...])                     # reduce over (N, *)
sigma_c = sqrt(var(x[:, c, ...]) + eps)          # reduce over (N, *)
x_hat   = (x - mu_c) / sigma_c                  # elementwise normalize
y       = gamma * x_hat + beta                   # scale and shift

# Update running statistics:
running_mean = (1 - momentum) * running_mean + momentum * mu_c
running_var  = (1 - momentum) * running_var  + momentum * var_c
```

**Forward pass (inference mode):**

```
y = gamma * (x - running_mean) / sqrt(running_var + eps) + beta
```

**Backward pass:**

```
# Gradients w.r.t. learnable parameters:
d_gamma = sum(grad_out * x_hat, dim=(N, *))
d_beta  = sum(grad_out,         dim=(N, *))

# Gradient w.r.t. input (requires mean/var from forward):
dx_hat  = grad_out * gamma
d_var   = sum(dx_hat * (x - mu) * -0.5 * (var + eps)^{-3/2}, dim=(N, *))
d_mu    = sum(dx_hat * -1/sigma + d_var * mean(-2*(x-mu)), dim=(N, *))
d_x     = dx_hat / sigma + d_var * 2*(x - mu)/m + d_mu/m
```

---

## Proposed Implementation

Following the TileOPs 2-layer architecture:

### Layer 1 — TileLang Kernels (`tileops/kernels/batch_norm/`)

| File | Description |
|------|-------------|
| `batch_norm_fwd.py` | Forward kernel: online Welford reduction + normalize + write running stats |
| `batch_norm_bwd.py` | Backward kernel: two-pass reduction for `d_gamma`, `d_beta`, and `d_x` |

**Key design decisions:**
- Use **Welford's online algorithm** for numerically stable single-pass mean+variance computation
- **Vectorized loads** (e.g., 128-bit) along the spatial/feature dimension
- **Shared memory reduction** across warps within each CTA
- Separate **training** and **inference** code paths (inference path can fuse into a single elementwise pass)
- Support `float16` / `bfloat16` inputs with `float32` accumulation for numerical stability

### Layer 2 — PyTorch Op (`tileops/ops/batch_norm.py`)

```python
class BatchNormFwdOp(Op):
    """Stateless batch normalization forward operator."""

    def forward(self, x, weight, bias, running_mean, running_var,
                training, momentum, eps) -> Tensor:
        ...

class BatchNormBwdOp(Op):
    """Stateless batch normalization backward operator."""

    def forward(self, grad_out, x, weight, mean, rstd) -> Tuple[Tensor, Tensor, Tensor]:
        ...
```

### Tests and Benchmarks

- `tests/ops/test_batch_norm.py` — correctness vs `torch.nn.functional.batch_norm`
- `benchmarks/ops/bench_batch_norm.py` — latency and bandwidth vs PyTorch (cuDNN backend)

**Test shapes to cover:**

| Variant | Shape |
|---------|-------|
| BatchNorm1d | `(N=32, C=512)`, `(N=32, C=512, L=1024)` |
| BatchNorm2d | `(N=32, C=256, H=56, W=56)` (typical ResNet) |
| BatchNorm3d | `(N=8, C=64, D=16, H=32, W=32)` |

---

## Expected Impact

Batch Normalization is a compute-bound and memory-bandwidth-bound operation. A fused TileLang kernel avoids the multi-kernel dispatch overhead of cuDNN and enables:

- **Fusion opportunities**: future `Conv + BN + ReLU` or `Linear + BN` fused kernels
- **Custom quantization-aware BN**: e.g., BN with FP8 accumulation for modern inference pipelines
- **Foundation for LayerNorm / GroupNorm variants** following the same reduction pattern

---

## Next Steps

- [ ] Implement `batch_norm_fwd.py` kernel with Welford reduction (training mode first)
- [ ] Implement `batch_norm_bwd.py` kernel
- [ ] Add inference-mode fast path (single elementwise pass)
- [ ] Write `BatchNormFwdOp` and `BatchNormBwdOp` wrappers
- [ ] Add correctness tests against `torch.nn.functional.batch_norm`
- [ ] Add benchmarks comparing against cuDNN baseline

A PR will follow after this issue is confirmed.

### PR DESCRIPTION
**Pull request by michaelwithu**
## Summary

Closes #303

Implements batch normalization forward and backward operators using TileLang kernels, following the standard TileOPs kernel→op delivery path.

### What's added

- **Kernels** (`tileops/kernels/batch_norm/`)
  - `BatchNormFwdTrainKernel`: two-pass training forward — computes batch mean/variance, normalizes, updates running stats, saves `mean`/`rstd` for backward
  - `BatchNormFwdInferKernel`: single-pass inference forward — fused scale/shift using precomputed running stats
  - `BatchNormBwdKernel`: two-pass backward — computes `grad_x`, `grad_weight`, `grad_bias` from saved `mean`/`rstd`

- **Op** (`tileops/ops/batch_norm.py`)
  - `BatchNormFwdOp`: wraps train + infer kernels; handles `(N, C, *spatial) → (C, L)` reshape internally; API mirrors `torch.nn.functional.batch_norm`
  - `BatchNormBwdOp`: wraps backward kernel with same reshape convention

- **Tests** (`tests/ops/test_batch_norm.py`): correctness validation against `torch.nn.functional.batch_norm` and `torch.autograd` for BatchNorm1d / BatchNorm2d, fp16 / bf16, training / inference modes

- **Benchmarks** (`benchmarks/ops/bench_batch_norm.py`): side-by-side comparison vs PyTorch cuDNN on ResNet-style shapes

### Performance design

Both fwd and bwd kernels use a **persistent path** when `L ≤ 8192`: data is loaded into shared memory once and reused for pass 2, eliminating the second global read. For larger `L`, a standard two-pass approach is used with non-power-of-2 `block_l` support to handle shapes like 3136 (= 56×56) with better occupancy.

Supported architectures: sm80, sm89, sm90.

## Benchmark Results

**Environment:** GPU: NVIDIA H200 · CUDA 12.8 · cuDNN 9.10.2 · PyTorch 2.9.1+cu128 · dtype: float16 · `tune=False` (default config, no autotuning)

**Forward (training mode) — TileOps vs torch cuDNN**

| N | C | spatial | TileOps (ms) | cuDNN (ms) | speedup |
|---|---|---------|-------------|------------|---------|
| 32 | 64 | — | 0.04 | 0.05 | 1.3x |
| 8 | 64 | (32,32) | 0.07 | 0.04 | 0.6x |
| 4 | 128 | (32,32) | 0.10 | 0.09 | 0.9x |
| 4 | 256 | (28,28) | 0.05 | 0.04 | 0.8x |
| 4 | 128 | (1024,1024) | 25.55 | 23.38 | 0.9x |
| 4 | 256 | (1024,1024) | 45.81 | 45.66 | 1.0x |

**Backward — TileOps vs torch autograd**

| N | C | spatial | TileOps (ms) | autograd (ms) | speedup |
|---|---|---------|-------------|--------------|---------|
| 32 | 64 | — | 0.01 | 0.12 | 12x |
| 8 | 64 | (32,32) | 0.11 | 0.16 | 1.5x |
| 4 | 128 | (32,32) | 0.11 | 0.09 | 0.8x |
| 4 | 256 | (28,28) | 0.06 | 0.18 | 3x |
| 4 | 128 | (1024,1024) | 34.35 | 67.83 | 2x |
| 4 | 256 | (1024,1024) | 47.01 | 102.54 | 2.2x |

**Performance AC (2026-03-06):** Backward 2.0–2.2x faster than torch autograd on large spatial (1024×1024); up to 12x on small shapes. Forward within 10% of cuDNN on large spatial; comparable on all other shapes. Accepted as functional parity at this stage (`tune=False`).

## Test plan

- [x] `conda run -n tileops python -m pytest tests/ops/test_batch_norm.py -vvs`
- [x] `conda run -n tileops python -m pytest benchmarks/ops/bench_batch_norm.py -vvs`
- [x] pre-commit passes on all changed files

### CODE REVIEW

File: tileops/kernels/batch_norm/batch_norm_fwd.py [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+            bl = threads * k
+            if bl >= L:
+                continue
+            if L % bl == 0:
+                return {"block_l": bl, "num_stages": 2, "threads": threads}
+    # Fallback (should rarely be reached).
+    for bl in [512, 256, 128, 64, 32, 16]:
+        if L % bl == 0:
+            return {"block_l": bl, "num_stages": 1, "threads": min(256, bl)}
+    return {"block_l": 16, "num_stages": 1, "threads": 16}
```
[the remark this answers is not in the document]
Comment (michaelwithu) [reply tileops/kernels/batch_norm/batch_norm_fwd.py]: Good catch. The unconditional return {"block_l": 16, ...} at the end of _find_best_block_l could silently produce wrong results when L % 16 != 0 — the kernel loop iterates L // block_l tiles and skips the remainder, leaving those positions in the torch.empty output uninitialized.

Fixed by replacing the silent fallback with a raise ValueError so the failure is loud and immediate rather than producing incorrect numerical output.

File: tileops/kernels/batch_norm/batch_norm_bwd.py [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+        for k in range(512 // threads, 0, -1):
+            bl = threads * k
+            if bl >= L:
+                continue
+            if L % bl == 0:
+                return {"block_l": bl, "num_stages": 2, "threads": threads}
+    for bl in [512, 256, 128, 64, 32, 16]:
+        if L % bl == 0:
+            return {"block_l": bl, "num_stages": 1, "threads": min(256, bl)}
+    return {"block_l": 16, "num_stages": 1, "threads": 16}
```
[the remark this answers is not in the document]
Comment (michaelwithu) [reply tileops/kernels/batch_norm/batch_norm_bwd.py]: Same issue as the forward kernel. Fixed identically — the fallback now raises ValueError instead of returning a block_l that does not divide L.

File: tileops/ops/batch_norm.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
+"""Batch Normalization Op.
+
+Wraps BatchNormFwdTrainKernel, BatchNormFwdInferKernel, and BatchNormBwdKernel
+in a standard TileOPs Op interface.
+
+User-facing API mirrors torch.nn.functional.batch_norm:
+
+    fwd_op = BatchNormFwdOp(N, C, *spatial, dtype=dtype)
+    y, mean, rstd = fwd_op(x, weight, bias, running_mean, running_var,
+                           training=True, momentum=0.1, eps=1e-5)
```
[the remark this answers is not in the document]
Comment (michaelwithu) [reply tileops/ops/batch_norm.py]: Fixed. The module-level docstring example now passes momentum and eps at construction time (BatchNormFwdOp(..., momentum=0.1, eps=1e-5)) rather than at the call site, which accurately reflects the actual API.

File: tileops/ops/batch_norm.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    def forward(
+        self,
+        x: torch.Tensor,
+        weight: torch.Tensor,
+        bias: torch.Tensor,
+        running_mean: torch.Tensor,
+        running_var: torch.Tensor,
+        training: bool = True,
+        momentum: Optional[float] = None,
+        eps: Optional[float] = None,
```
[the remark this answers is not in the document]
Comment (michaelwithu) [reply tileops/ops/batch_norm.py]: Fixed. Both parameters have been removed from forward. They were accepted but never read — their values are baked into the kernel at __init__ time via self.momentum and self.eps. Keeping them in the signature was misleading and risked callers believing they could override these values per-call.

File: benchmarks/ops/bench_batch_norm.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+"""
+
+import math
+from typing import Optional
+
+import pytest
+import torch
+
+from benchmarks.benchmark import BenchmarkBase, BenchmarkReport
+from tests.ops.test_batch_norm import BatchNormFwdFixture, BatchNormBwdFixture
```
[the remark this answers is not in the document]
Comment (michaelwithu) [reply benchmarks/ops/bench_batch_norm.py]: Fixed. Replaced the unused BatchNormFwdFixture / BatchNormBwdFixture top-level imports with BatchNormFwdTest / BatchNormBwdTest, which are the symbols actually used in the benchmark functions

File: .claude/settings.json [hunk 6]
```diff
@@ -1,4 +1,20 @@
...
+      "Read(//home/jieneng.yu/workspace/TileOPs/**)",
+      "Bash(conda run -n tileops python -m pytest tests/ops/test_batch_norm.py -vvs -x)",
+      "Bash(conda run -n tileops python -m pytest tests/ops/test_batch_norm.py -vvs)",
+      "Bash(conda run -n tileops python -m pytest benchmarks/ops/bench_batch_norm.py -vvs)",
+      "Bash(pytest tests/ops/test_batch_norm.py -vvs)",
+      "Bash(conda run -n tileops python -c \":*)",
+      "Bash(conda run -n tileops python -m pytest benchmarks/ops/bench_batch_norm.py -vvs --timeout=120)",
+      "Bash(conda run -n tileops python -m pytest \"benchmarks/ops/bench_batch_norm.py::test_batch_norm_fwd_bench[8-64-spatial1-dtype1-True-False]\" -vvs)",
+      "Bash(conda run -n tileops python -m pytest benchmarks/ops/bench_batch_norm.py -vvs 2)",
+      "Bash(1 echo \"Exit code: $?\" cat /tmp/bench_out.txt)"
```
Comment (lcy-seso) [inline .claude/settings.json]: Developer-local permissions with hardcoded path (`/home/jieneng.yu/...`). Should not be committed — revert this file.

File: tileops/kernels/batch_norm/batch_norm_fwd.py [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+                    var_val + T.cast(eps, accum_dtype))
+
+                # Save for backward.
+                mean_out[bc] = mean_val
+                rstd_out[bc] = rstd_val
+
+                # Update running statistics.
+                mom = T.cast(momentum, accum_dtype)
+                running_mean[bc] = (T.cast(1.0, accum_dtype) - mom) * running_mean[bc] + mom * mean_val
+                running_var[bc] = (T.cast(1.0, accum_dtype) - mom) * running_var[bc] + mom * var_val
```
Comment (lcy-seso) [inline tileops/kernels/batch_norm/batch_norm_fwd.py]: `running_var` updated with biased variance (`var_val = sum(x^2)/L - mean^2`). PyTorch uses Bessel correction: `unbiased_var = var_val * L / (L - 1)`. This causes numerical drift from PyTorch over many forward passes.

**Review by lcy-seso — COMMENTED** [review]
Architecture is clean — 2-layer separation, persistent/non-persistent path design is solid. Three issues:

1. `.claude/settings.json` has developer-local permissions (hardcoded home path). Should not be in this PR.
2. Agree with @gemini-code-assist: `running_var` uses biased variance. PyTorch applies Bessel correction (`L/(L-1) * var`). Will drift over many batches.
3. Tests only validate output `y`, not `running_mean`/`running_var` — the biased variance bug is untested.

See inline comments.

File: tileops/kernels/batch_norm/batch_norm_fwd.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (michaelwithu) [reply tileops/kernels/batch_norm/batch_norm_fwd.py]: Fixed. Added Bessel's correction: `unbiased_var = var_val * L / (L - 1)` and used it for the running_var update instead of the biased `var_val`.

File: tests/ops/test_batch_norm.py [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+    y, mean, rstd = op(x, weight, bias, running_mean, running_var, training=training)
+
+    ref_y, *_ = test.ref_program(x, weight, bias, running_mean, running_var)
+
+    # float16 accumulates more error; use loose tolerances.
+    atol, rtol = (1e-2, 1e-2) if dtype == torch.float16 else (2e-2, 2e-2)
+    max_err = (y.float() - ref_y.float()).abs().max()
+    assert torch.allclose(y.float(), ref_y.float(), atol=atol, rtol=rtol), \
+        f"fwd mismatch (training={training}): max_err={max_err:.4e}"
+    print(f"test_batch_norm_fwd passed [training={training}]: max_err={max_err:.4e}")
```
[the remark this answers is not in the document]
Comment (michaelwithu) [reply tests/ops/test_batch_norm.py:166]: Fixed. Extended `test_batch_norm_fwd` to clone running stats before the op call and assert `running_mean`/`running_var` match the PyTorch reference when `training=True`. Also added a non-persistent path test case (L=16384) to both fixtures.

File: .claude/settings.json (hunk 6 above)
Comment (michaelwithu) [reply .claude/settings.json]: Fixed. Reverted to the main-branch version — the developer-local `permissions` block has been removed.

File: tileops/kernels/batch_norm/batch_norm_fwd.py (hunk 7 above)
Comment (michaelwithu) [reply tileops/kernels/batch_norm/batch_norm_fwd.py]: Fixed. Used `unbiased_var = var_val * L / (L - 1)` (Bessel's correction) for the running_var update to match PyTorch's behavior.
